docs/kuneiform/procedures.md

```diff
@@ -0,0 +1,275 @@
+
+```ts
+procedure declare_vars() public {
+    // declare and assign new int var:
+    $var1 int := 1;
+
+    // declare a new var to be assigned to later
+    $var2 text;
+
+    // assign to $var2. type cast to match type text:
+    $var2 := $var1::text;
+}
+```
+
+### Arithmetic
+
+Basic arithmetic can be performed in procedures:
+
+```ts
+procedure do_math($num1 int, $num2 int) public {
+    // add and divide, with precedence:
+    $avg int := ($num1 + $num2)/2;
+
+    // modulo:
+    $mod int := $num2%10;
+
+    // negate
+    $neg_mul int := (-$avg)*$mod;
+}
+```
+
+### Executing SQL
+
+SQL statements can be executed at any point in a procedure. In the below example, we generated a deterministic UUID using a random UUID namespace and the transaction ID. We then use the generated id as the user's primary key.
+
+```ts
+procedure create_user($name text, $age int) public {
+    $id uuid := generate_uuid_v5('455f60aa-0569-4aaa-8469-63be2ec4dd96'::uuid, @txid)
+    // do something
+    // ...
+    INSERT INTO users (id, name, age)
+    VALUES ($id, $name, $age);
+}
+```
+
+### Handling Arrays
+
+Procedures support basic array operations. It supports the following functions:
+
+- `array_append(type[], type) type[]`: appends a value to the end of an array.
+- `array_prepend(type, type[]) type[]`: prepends a value to the beginning of an array.
+- `array_cat(type[], type[]) type[]`: concatenates two arrays together.
+- `array_length(type[]) int`: returns the length of the array.
+
+**Arrays are 1-indexed.**
+
+```ts
+procedure handle_array() public view returns (val text) {
+    $arr text[] := ['hello', 'world'];
+    $arr := array_append($arr, 'goodbye');
+
+    $len int := array_length($arr); // $len == 3
+
+    // return the first value.
+    return $arr[1];
+}
+```
+
+### If/Then
+
+```ts
+procedure age_conditional($age int) public {
+    if $age < 0 {
+        // do something
+    } elseif $age < 18 {
+        // do something
+    } else {
+        // something else
+    }
+}
+```
+
+### For Loop
+
+For loops can be declared over 4 different targets: arrays, ranges, procedure calls, and SQL statements.
+
+#### Array Loop
+
+```ts
+procedure iterate_over_array($names text[]) public {
+    for $name in $names {
+        // $name is of type `text`
+    }
+}
+```
+
+#### Range Loop
+
+```ts
+procedure iterate_over_range() public {
+    // iterating over all numbers from 1 to 25
+    for $i in 1..25 {
+        // $i is type int
+    }
+}
+
+procedure iterate_range_dynamic($start int, $end int) public {
+    // can also iterate over dynamic ranges
+    for $i in $start..$end {
+        // ...
+    }
+}
+```
+
+#### Procedure Loops
+
+You can iterate over any SQL statement as well. The following examples reference an example table `users`, with columns `id`, `name`, and `age`.
+
+```ts
+procedure iterate_over_sql() public {
+    for $row in SELECT id, name, age from users {
```

Comment:
SQL formatting is a little inconsistent.  `SELECT` is caps, but `from` is not.  Would look better in caps.

In the subsequent example with `iterate_over_delete` it's all lower case (`delete from users ...`), which imo makes it harder to read when it follows `in`.

Then in `iterate_over_users` it looks proper with `for $row in SELECT * FROM get_users() {`